src/priceClient/priceClient.ts

```diff
@@ -0,0 +1,177 @@
+      if (Object.keys(prices).length === 0) {
+        this.logger.warn({ at: "PriceClient#getPricesByAddress", message: "Failed to update token prices." });
+        // @todo throw ?
+        return [];
+      }
+      this.updateCache(priceCache, prices, requestAddresses);
+    }
+
+    return addresses.map((addr: string) => {
+      const { price, timestamp } = priceCache[addr.toLowerCase()];
+      return { address: addr, price, timestamp };
+    });
+  }
+
+  expireCache(currency: string): void {
+    const priceCache = this.getPriceCache(currency);
+    Object.values(priceCache).forEach((token: TokenPrice) => (token.timestamp = 0));
+    this.logger.debug({ at: "PriceClient#expireCache", message: `Expired ${currency} cache.` });
+  }
+
+  protected getPriceCache(currency: string): PriceCache {
+    if (this.prices[currency] === undefined) this.prices[currency] = {};
+    return this.prices[currency];
+  }
+
+  private async requestPrices(addresses: string[], currency: string): Promise<PriceCache> {
+    let prices: TokenPrice[] = [];
+
+    for (const priceFeed of this.priceFeeds) {
+      this.logger.debug({
+        at: "PriceClient#getPriceByAddress",
+        message: `Looking up prices via ${priceFeed.name}.`,
+      });
+      try {
+        prices = await priceFeed.getPricesByAddress(addresses, currency);
+        if (prices.length === 0) {
+          throw Error("Zero-length response received");
+        }
+      } catch (err) {
+        this.logger.debug({
+          at: "PriceClient#requestPrices",
+          message: `Price lookup against ${priceFeed.name} failed (${err}).`,
+          tokens: addresses,
+        });
+        continue; // Failover to the next price feed...
+      }
+    }
+
+    if (prices.length === 0) {
+      throw Error(`Price lookup failed against all price feeds (${this.listPriceFeeds().toString()})`);
+    }
+
+    return Object.fromEntries(prices.map((price) => [price.address.toLowerCase(), price]));
+  }
+
+  private updateCache(priceCache: PriceCache, prices: PriceCache, expected: string[]): void {
+    const updated: TokenPrice[] = [];
+    const skipped: { [token: string]: string } = {}; // Includes reason for skipping
+    const now: number = msToS(Date.now());
+
+    expected.forEach((address: string) => {
+      const addr = address.toLowerCase(); // for internal priceCache lookup.
+      const tokenPrice: TokenPrice | undefined = prices[addr];
+
+      if (tokenPrice === undefined) {
+        skipped[address] = "Not included in price feed response.";
+      } else if (tokenPrice.timestamp > now) {
+        skipped[address] = `Token price timestamp is too new (timestamp ${tokenPrice.timestamp}).`;
+      } else if (tokenPrice.timestamp >= priceCache[addr].timestamp) {
+        const { price, timestamp } = tokenPrice;
+        // @todo: Do we care if the token price is older than maxPriceAge?
+        priceCache[addr] = { address: "unused", price: price, timestamp: timestamp };
```

Comment:
Why are we assigning the address field "unused" for a real value? Why does this field exist?

Reply:
> Why are we assigning the address field "unused" for a real value? Why does this field exist?

All our existing code is built around a CoinGeckoPrice type, which follows this type definition:
```
export type CoinGeckoPrice = {
  address: string;
  timestamp: number;
  price: number;
};
```
I'm reusing that type for convenience/simplicity in the code, but don't have any need for the address member because I map the prices by their address, instead of storing them in a list or set. I prefer this since I don't like searching each time I need to retrieve a price.

So...the address member is superfluous to my needs, but it means I don't need to define a new type that has only one difference to the existing CoinGeckoPrice type that we already depend on.

Reply:
Oh, I would use a different type in this situation! You don't have to change in this PR, but I think if that field is just not being used, people will extract it and think that it has some purpose or meaning.